cmd/juju/charmhub/download.go

```diff
@@ -0,0 +1,221 @@
+
+	"github.com/juju/juju/api/modelconfig"
+	"github.com/juju/juju/charmhub"
+	"github.com/juju/juju/charmhub/transport"
+	jujucmd "github.com/juju/juju/cmd"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/environs/config"
+)
+
+const (
+	downloadSummary = "Locates and then downloads a CharmHub charm."
+	downloadDoc     = `
+Download a charm from the CharmHub store by a specified name.
+
+Examples:
+    juju download postgresql
+
+See also:
+	info
+	find
+`
+)
+
+// NewDownloadCommand wraps downloadCommand with sane model settings.
+func NewDownloadCommand() cmd.Command {
+	return modelcmd.Wrap(&downloadCommand{})
+}
+
+// downloadCommand supplies the "download" CLI command used for downloading
+// charm snaps.
+type downloadCommand struct {
+	modelcmd.ModelCommandBase
+	out cmd.Output
+
+	modelConfigAPI ModelConfigGetter
+	charmHubClient CharmHubClient
+
+	charmHubURL   string
+	charmOrBundle string
+	archivePath   string
+}
+
+// Info returns help related download about the command, it implements
+// part of the cmd.Command interface.
+func (c *downloadCommand) Info() *cmd.Info {
+	download := &cmd.Info{
+		Name:    "download",
+		Args:    "[options] <charm>",
+		Purpose: downloadSummary,
+		Doc:     downloadDoc,
+	}
+	return jujucmd.Info(download)
+}
+
+// SetFlags defines flags which can be used with the download command.
+// It implements part of the cmd.Command interface.
+func (c *downloadCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ModelCommandBase.SetFlags(f)
+	f.StringVar(&c.charmHubURL, "charm-hub-url", "", "specify url for querying the charmhub store")
+	f.StringVar(&c.archivePath, "path", "", "path to download the charm to")
+}
+
+// Init initializes the download command, including validating the provided
+// flags. It implements part of the cmd.Command interface.
+func (c *downloadCommand) Init(args []string) error {
+	if len(args) != 1 {
+		return errors.Errorf("expected a charm or bundle name")
+	}
+	if err := c.validateCharmOrBundle(args[0]); err != nil {
+		return errors.Trace(err)
+	}
+	c.charmOrBundle = args[0]
+	return nil
+}
+
+func (c *downloadCommand) validateCharmOrBundle(charmOrBundle string) error {
+	curl, err := charm.ParseURL(charmOrBundle)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !charm.CharmHub.Matches(curl.Schema) {
+		return errors.Errorf("%q is not a Charm Hub charm", charmOrBundle)
+	}
+	return nil
+}
+
+// Run is the business logic of the download command.  It implements the meaty
+// part of the cmd.Command interface.
+func (c *downloadCommand) Run(cmdContext *cmd.Context) error {
+	var charmHubURL string
+	if c.charmHubURL != "" {
+		charmHubURL = c.charmHubURL
+	} else {
+		config, err := c.getModelConfig()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		charmHubURL, _ = config.CharmHubURL()
+	}
+
+	config, err := charmhub.CharmHubConfigFromURL(charmHubURL, downloadLogger{
+		Context: cmdContext,
+	})
+	if err != nil {
+		return errors.Trace(err)
```

Comment:
Add user friendly text to this and some of the other errors:
$ juju download ubuntu-qa --charm-hub-url "heather"
ERROR Get "heather/v2/charms/info/ubuntu-qa": unsupported protocol scheme ""